rfcs/049-catalogue-api-aggregations-modelling.md

```diff
@@ -0,0 +1,176 @@
+        "data": {
+          "id": "eng",
+          "label": "English",
+          "type": "Language"
+        },
+        "count": 691840,
+        "type": "AggregationBucket"
+      },
+      {
+        "data": {
+          "id": "fre",
+          "label": "French",
+          "type": "Language"
+        },
+        "count": 67187,
+        "type": "AggregationBucket"
+      },
+...
+```
+
+This presents a clean, consistent interface to clients – a value looks the same whether it's in a work or in an aggregation.
+
+We don't to change this behaviour.
+
+
+
+## The current implementation, and the problems it poses
+
+The API uses [Elasticsearch aggregations][es_aggs] to aggregate over a field in internal model.
+Elasticsearch will return single string values, which the API then serialises into the display model.
+
+For example, our language aggregation starts as [an ES terms aggregation][terms_agg] over the `data.languages.id` field:
+
+```http
+GET /works-indexed-2022-04-28/_search
+{
+  "aggs": {
+    "languages": {
+      "terms": { "field": "data.languages.id" }
+    }
+  }
+}
+```
+
+The Elasticsearch API response can only use a single string in its buckets, for example:
+
+```
+{
+  "aggregations" : {
+    "languages" : {
+      "buckets" : [
+        {
+          "key" : "eng",
+          "doc_count" : 691840
+        },
+        {
+          "key" : "fre",
+          "doc_count" : 67187
+        },
+        ...
+```
+
+The API has to contain enough internal/display model logic to interpret these values -- to know that, say, `eng` means English and it's serialised as id/label/type.
+This is precisely the sort of model coupling we're trying to get away from.
+
+For more complex types, we've had to jump through hoops to shoehorn aggregations into this approach -- e.g. contributor values are stored like `person:Henry Wellcome` because we need both the type and the label to return them correctly.
+
+It would be nice if we could remove this coupling and simplify how aggregations work in the API.
+
+[es_aggs]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html
+[terms_agg]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html
+
+
+
+## Proposed solution
+
+We add a new field `query.aggregatableValues` to the Elasticsearch model.
+This would be an object field in which the values are lists of strings:
+
+```
+type Query {
+  aggregatableValues: Map[String, List[String]]
+  ...
+}
+```
+
+The labels are the aggregation types (languages, work types, licenses) and the values are blobs of display JSON stored as strings.
+That is, we create the display JSON, then serialise the whole object as a string, and store that.
+
+This is easiest to understand with an example:
+
+```json
+{
+  "id": "example-work",
+  "query": {
+    "aggregatableValues": {
+      "languages": [
+        " { \"id\" : \"eng\", \"label\": \"English\", \"type\": \"Language\" } ",
```

Comment:
Are there scenarios where the type value here might not be the same for all aggregations of these values?

Reply:
Yes, contributors can be Agent/Person/Organisation.

Reply:
e.g. people - might we create aggregations where we look for people in both subject and contributor fields, to create an aggregation called "people", but also one where we look in just the contributor field to find "contributors".

Reply:
I can't think of a scenario where we'd aggregate across fields like you're describing.

Reply:
The reason for the question is about whether there's any situation where having type in here would be problematic.  I doubt if it would be.

It looks like the type will always be just the type of each object within the aggregation, and not have any relation to the type of the aggregation being performed.